edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsWorker.cs

```diff
@@ -0,0 +1,152 @@
+            this.uploader = uploader;
+            this.logger = logger;
+            this.systemTime = systemTime ?? SystemTime.Instance;
+        }
+
+        public void Start(TimeSpan scrapingInterval, TimeSpan uploadInterval)
+        {
+            this.scrape = new PeriodicTask(this.Scrape, scrapingInterval, scrapingInterval, this.logger, "Scrape");
+            this.upload = new PeriodicTask(this.Upload, uploadInterval, uploadInterval, this.logger, "Scrape");
+        }
+
+        async Task Scrape(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.GetLock(cancellationToken))
+            {
+                this.logger.LogInformation("Scraping Metrics");
+                List<Metric> metricsToPersist = new List<Metric>();
+
+                foreach (var moduleResult in await this.scraper.ScrapeAsync(cancellationToken))
+                {
+                    IEnumerable<Metric> scrapedMetrics = MetricsParser.ParseMetrics(this.systemTime.UtcNow, moduleResult.Value);
+
+                    foreach (Metric scrapedMetric in scrapedMetrics)
+                    {
+                        // Get the previous scrape for this metric
+                        if (this.metrics.TryGetValue(scrapedMetric.GetValuelessHash(), out Metric oldMetric))
+                        {
+                            // If the metric is unchanged, do nothing
+                            if (oldMetric.Value.Equals(scrapedMetric.Value))
+                            {
+                                continue;
+                            }
+
+                            // if the metric has changed, write the previous metric to disk
+                            metricsToPersist.Add(oldMetric);
+                        }
+
+                        // if new metric or metric changed, save to local buffer
+                        this.metrics[scrapedMetric.GetValuelessHash()] = scrapedMetric;
+                    }
+                }
+
+                this.logger.LogInformation("Scraped Metrics");
+
+                if (metricsToPersist.Count != 0)
+                {
+                    this.logger.LogInformation("Storing Metrics");
+                    this.storage.AddScrapeResult(Newtonsoft.Json.JsonConvert.SerializeObject(metricsToPersist));
+                    this.logger.LogInformation("Stored Metrics");
+                }
+            }
+        }
+
+        async Task Upload(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.GetLock(cancellationToken))
+            {
+                this.logger.LogInformation("Uploading Metrics");
+                await this.uploader.UploadAsync(this.GetMetricsToUpload(this.lastUploadTime), cancellationToken);
+                this.lastUploadTime = this.systemTime.UtcNow;
+                this.logger.LogInformation("Uploaded Metrics");
+            }
+        }
+
+        IEnumerable<Metric> GetMetricsToUpload(DateTime lastUploadTime)
```

Comment:
Can you comment this function? I see that metrics are coming from two places, but I would like documented why they come from two places.

Reply:
added some comments. the most relevant one added is 
```
// This acts as a local buffer. It stores the previous value of every metric. If the new value for that metric is unchanged,
// it doesn't write the duplicate value to disk.
Dictionary<int, Metric> metrics = new Dictionary<int, Metric>();
```